NIFI-12372 MiNiFi C2 Encrypt Flow Configuration Properties when Transferring

briansolo1985 opened the pull request:
<!-- Licensed to the Apache Software Foundation (ASF) under one or more -->
<!-- contributor license agreements.  See the NOTICE file distributed with -->
<!-- this work for additional information regarding copyright ownership. -->
<!-- The ASF licenses this file to You under the Apache License, Version 2.0 -->
<!-- (the "License"); you may not use this file except in compliance with -->
<!-- the License.  You may obtain a copy of the License at -->
<!--     http://www.apache.org/licenses/LICENSE-2.0 -->
<!-- Unless required by applicable law or agreed to in writing, software -->
<!-- distributed under the License is distributed on an "AS IS" BASIS, -->
<!-- WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. -->
<!-- See the License for the specific language governing permissions and -->
<!-- limitations under the License. -->

# Summary

[NIFI-12372](https://issues.apache.org/jira/browse/NIFI-12372)

# Tracking

Please complete the following tracking steps prior to pull request creation.

### Issue Tracking

- [x] [Apache NiFi Jira](https://issues.apache.org/jira/browse/NIFI) issue created

### Pull Request Tracking

- [x] Pull Request title starts with Apache NiFi Jira issue number, such as `NIFI-00000`
- [x] Pull Request commit message starts with Apache NiFi Jira issue number, as such `NIFI-00000`

### Pull Request Formatting

- [x] Pull Request based on current revision of the `main` branch
- [x] Pull Request refers to a feature branch with one commit containing changes

# Verification

Please indicate the verification steps performed prior to pull request creation.

### Build

- [x] Build completed using `mvn clean install -P contrib-check`
  - [x] JDK 21

### Licensing

- [ ] New dependencies are compatible with the [Apache License 2.0](https://apache.org/licenses/LICENSE-2.0) according to the [License Policy](https://www.apache.org/legal/resolved.html)
- [ ] New dependencies are documented in applicable `LICENSE` and `NOTICE` files

### Documentation

- [ ] Documentation formatting appears as expected in rendered files

briansolo1985 commented:
A flow incoming via C2 may contain unencrypted sensitive properties as the agents can have there our encryption key.
The goal is to check the properties in the incoming flow, and if there are unencrypted sensitive properties there, use the configured encryption key to encrypt them before persisting the flow.json.raw to disk.
The logic first tries to utilise the property descriptors sent with the flow to detect the sensitive property names. However the property descriptors can omitted. In this case the logic uses the runtime manifest to get the list of sensitive properties.

exceptionfactory commented on minifi/minifi-nar-bundles/minifi-framework-bundle/minifi-framework/minifi-framework-core/src/main/java/org/apache/nifi/minifi/c2/command/FlowPropertyEncryptor.java:
```diff
@@ -0,0 +1,167 @@
+import org.apache.nifi.flow.VersionedPropertyDescriptor;
+import org.apache.nifi.minifi.commons.service.FlowSerDeService;
+
+public class FlowPropertyEncryptor {
```
I recommend defining an interface with a Standard implementation.

briansolo1985 replied:
Introduced the interface

exceptionfactory commented on minifi/minifi-nar-bundles/minifi-framework-bundle/minifi-framework/minifi-framework-core/src/test/java/org/apache/nifi/minifi/c2/command/FlowPropertyEncryptorTest.java:
```diff
@@ -0,0 +1,338 @@
+
+    @Test
+    public void shouldEncryptParameterContextsSensitiveVariables() {
+        // given
```
The `given/when/then` comments should be removed as they are not standard across the repository and do not necessarily add much to the test definition.

briansolo1985 replied:
Removed the comments

exceptionfactory reviewed (commented):
Thanks for working on this improvement @briansolo1985.

The general approach makes sense, and I will defer to others with more familiarity for closer review. The use of the standard PropertyEncryptor looks good.

I recommend defining an interface named `FlowPropertyEncryptor` with a `StandardFlowPropertyEncryptor` implementation, particularly since the test class mocks the class right now.

bejancsaba reviewed (approved):
@briansolo1985 thank you for this enhancement the code and tests are clean and straightforward. I tried it via C2 and the downloaded raw flow that is persisted contained the encrypted property on the file system while the decrypted value was present runtime. So works as expected. +1 from my side.

ferencerdei commented:
Thanks for the feature, this really useful. I have only one question.

briansolo1985 commented on minifi/minifi-nar-bundles/minifi-framework-bundle/minifi-framework/minifi-framework-core/src/main/java/org/apache/nifi/minifi/c2/command/DefaultUpdateConfigurationStrategy.java:
```diff
@@ -90,11 +93,12 @@ public boolean update(byte[] rawFlow) {
 
         try {
-            byte[] enrichedFlowCandidate = flowEnrichService.enrichFlow(rawFlow);
+            byte[] propertyEncryptedRawDataFlow = flowPropertyEncryptor.encryptSensitiveProperties(rawFlow);
```
We only need to encrypt the raw flow.json. Sensitive properties in the enriched flow are automatically encrypted when loaded by flowService.

briansolo1985 commented:
Thanks your all the reviews. Addressed the comments, and pushed the most recent changes. Please re-review.

exceptionfactory commented on minifi/minifi-nar-bundles/minifi-framework-bundle/minifi-framework/minifi-framework-core/src/test/java/org/apache/nifi/minifi/c2/command/FlowPropertyEncryptorTest.java:
```diff
@@ -0,0 +1,329 @@
+import org.mockito.junit.jupiter.MockitoExtension;
+
+@ExtendWith(MockitoExtension.class)
+public class FlowPropertyEncryptorTest {
```
This class should be renamed to `StandardFlowPropertyEncryptorTest` to match the new implementation class name.

briansolo1985 replied:
Thanks for the catch, renamed.

exceptionfactory reviewed (commented):
Thanks for defining the interface @briansolo1985, that is helpful.

The interface definition itself is very generic, and presumes that the implementation handles parsing and serializing the JSON. Are there any performance concerns related to flow size? It looks like the enrichment service expects the flow as a byte array, but should this change also prompt further adjustments to avoid multiple rounds for conversion from byte array to JSON objects and back again?

briansolo1985 commented:
I addressed the second round of comments. Decided to make a bigger refactor: moved services to minifi-commons, created interfaces for those services to become consistent. Also removed the flow serialization and deserialization from those particular services. Performance was not an issue there as that code was not on a critical path, eg flow updates are expected maximum a few times a day, but most frequently we talk about weeks or months here. However having the serde in the services hurt SRP at some level, so moved the logic to an upper layer. Also now the flow serde logic is encapsulated in one class.
Rebased onto the latest main. Tested a flow with C2, unecrypted properties were encrypted correctly.
Please review the latest changes.

ferencerdei reviewed (approved):
Thanks for the updates. +1 from my side.
I'm happy to merge if there is no other objection.
